subsys/net/lib/http/http_client.c

```diff
@@ -408,49 +409,63 @@ static void http_client_init_parser(struct http_parser *parser,
 	settings->on_url = on_url;
 }
 
-static int http_wait_data(int sock, struct http_request *req)
+static int http_wait_data(int sock, struct http_request *req, int32_t timeout)
 {
 	int total_received = 0;
 	size_t offset = 0;
 	int received, ret;
+	struct pollfd poll_fd;
+
+	poll_fd.fd = sock;
+	poll_fd.events = POLLIN;
+
+	ret = poll(&poll_fd, 1, timeout);
+	if (ret < 0) {
+		LOG_ERR("Socket poll err (%d)", errno);
+		return -errno;
+	} else if (ret == 0) {
+		LOG_ERR("Socket poll timeout");
+		return ret;
+	} else if (poll_fd.revents & POLLIN) {
+		do {
```

Comment:
The internal do/while loop doesn't seem right - `recv()` might've read all of the data available,  but still expect more, so it'll block anyway. A proper way would be to `poll()` at the beginning of each loop iteration. It'd also keep the diff smaller - no extra indentation.

Reply:
ah, you are right. if I read it correctly, you mean
```c
do{
  poll();
  if(ret<0){
    ret =  -errno;
    break;
  }else if(ret == 0){
    break;
  }
  recv();
  ...
  // break if message_complete or errno
} while(true)
```
right?

Reply:
Yeah, something like this. It'd also make sense to check if `POLLIN` was returned, because if not, it's likely an error condition (`POLLERR` etc).